src/what_is_an_ABI.md

```diff
@@ -32,3 +32,13 @@ To quote from the initial proposal:
 > TODO: more recent examples.
 
 ## What would a stable modular ABI look like for Rust?
+
+- **Fully specified** - There should be no ambiguous cases, either in how to map Rust to the ABI, or how the ABI maps back to Rust objects.
+- **Versioned** - There should be a mechanism to determine which version of the ABI a chunk of code is using by querying the code itself.  The version should follow [semver](https://semver.org/) conventions, and should be both machine and human parseable, so that linkers and loaders can decide if two pieces of code are compatible (from an ABI standpoint).  Under some circumstances, it may be possible for a linker or loader to generate code to bind chunks of code together that have different ABI versions.
+- **Designed for introspection** - There are tools for different operating systems to analyze blobs of code to determine what they do.  For Linux, there are [nm](https://linux.die.net/man/1/nm), [objdump](https://linux.die.net/man/1/objdump), and a number of other utilities.  Unfortunately, the tools are limited by the knowledge encoded in the executable formats; in most cases you can't read the docs for a library by reading the library itself.  A useful addition would be the ability to store the documentation for a library in one of the segments of the library, with tools that are able to find and display that information.
```

Comment:
I think that this is generally a good idea, but it's not specifically related to the ABI - while useful, embedding documentation inside a compiled executable seems to be a bit out-of-scope. I could see, however, binaries including ABI information so that linking compilers could automatically understand the API provided to link against.

Reply:
Sure.  I meant to make that an example, but got rushed (my son started playing with my keyboard while I was typing, which made things difficult).  

That said, even if the entire API isn't documented it would still be nice to have some metadata within the binary itself.  E.g., the [keys in `Cargo.toml`](https://doc.rust-lang.org/cargo/reference/manifest.html) would be the basis of the metadata.  That would allow us (as humans) to figure out where to go for more information (the documentation URL), as well as allow loaders to figure out *if* they can link a particular pair of binaries together (from the dependencies, which are versioned).